notes/fomega/evaluation/builtin-evaluation.md

```diff
@@ -0,0 +1,285 @@
+other variables as well: the new machine appeared attractive because it would
+reduce the number of irrelevant variables stored in environments).
+
+In James' original proposal, closures were to be replaced with a new kind
+of "value":
+
+```
+V := con cn
+   | abs ɑ (M , ⍴)
+   | lam x (M , ρ)
+   | iwrap A B V
+```
+
+These are similar to, but distinct from, the "values" in the Plutus Core
+specification, which are terms that can't be reduced any further.
+
+Roman's proposal was to change this to
+
+```
+V := con cn
+   | abs ɑ (M , ⍴)
+   | lam x (M , ρ)
+   | iwrap A B V
+   | builtin ρ bn count [types] [V]
+```
+(we'll call these things "CEK-values")
+
+The `builtin` constructor contains
+
+  * The name `bn` of the builtin
+  * The environment `ρ` in effect at the time that `(builtin bn)` was first encountered.
+  * A count of the arguments.  This is filled in with the arity of `bn`
+    (looked up in a table) when `(builtin b)` was first encountered
+  * A list of type arguments
+  * A list of term arguments
+
+The lists of arguments are initially empty: whenever a `builtin`
+CEK-value is instantiated or applied to an argument the argument is
+added to the appropriate list and the count is decremented.  When the
+count reaches zero the builtin name and the arguments can be passed
+directly to the constant application machinery.
+
+(This technique is slightly fragile because we only count the total
+number of type and term arguments, and don't check that type and term
+arguments are correctly interleaved.  However, (a) these _will_ be
+typed by the typechecker, (b) we'll eventually be moving to an untyped
+evaluator where this is less important, and (c) if we really wanted to
+be more robust we could keep separate lists of type and term arities,
+or (better, but at a slight cost) provide a list of `TypeArg |
+TermArg` values, removing the first element every time we encounter an
+instantiation or application, and checking that it is of the correct kind.)
+
+This process is clearly much more efficient than that in the original
+CEK machine, and in fact enabled the alternative CEK machine with
+unsaturated builtins to become competitive with the one with saturated
+builtins.
+
+
+### The CK machine
+
+Unfortunately we can't use the above technique with the CK machine:
+that works directly with values as described in the Plutus Core
+specification, and there's nowhere to store the argument information.
+I've considered a number of ways to fix this.
+
+ * I initially thought we could store the required information
+   in a frame.  This is how the CK machine worked with saturated
+   builtins, but we can't use the same technique with unsaturated
+   builtins.  In the unsaturated case a partially applied builtin
+   is a perfectly legitimate value and could be returned by a function
+   or supplied as an argument to another function.  This just doesn't
+   fit in with the way that stack frames are used.
```

Comment:
Let me think that through properly.  I'd managed to convince myself it didn't work, but now I'm having doubts again.  I think that proves that I should write something down...

Reply:
I think the difference is that a lambda does effectively carry its arity with it: it's always 1. So the frame for application knows what to do just from the function term if it's a lambda.

In our case we *could* reconstruct that: the application frame could deconstruct its function term (e.g. using `termAsPrimIterApp`, look up the corresponding arities, and decide whether it's done or not. But that would be repeating a lot of work again.

Reply:
Looks like what I said here is exactly what you said you were doing below!

Reply:
I think trying to use frames for partially-applied builtins in the CK machine really forces you to introduce a new type of value to convert the frame into when you run out of arguments and you're still not fully-applied, and this just takes us back to the version with special values. I've updated the document to reflect that.

Reply:
> I think trying to use frames for partially-applied builtins in the CK machine really forces you to introduce a new type of value to convert the frame into when you run out of arguments and you're still not fully-applied, and this just takes us back to the version with special values. 

Having

```
s, [(λf:int->int . f (con 1)), _], (builtin addInteger 1 [(con 11]) ▻ ???
```

can we turn the builtin frame into a normal term (`addInteger` applied to `11`) and continue evaluation _in the return mode_? Then we'll eventually need to retraverse the same application and turn the term into the frame once again when we have all arguments in place, but this inefficiency is probably not of a much concern, because a partially applied builtin fed to another function is not a very common case and it'll probably be faster than the current solution anyway.

tl;dr: temporary turn builtin frames into stuck applications (known to be values) when we need that and reprocess them later.

Reply:
I think the values as described in the plutus core spec are designed for use with saturated builtins. So, we should add a constructor for an unsaturated builtin to the ordinary definition of values. We should update the reduction semantics to use this too. Would it help to put the arity info in there?

I have an Agda implementation of an CK + unsaturated builtins: https://github.com/input-output-hk/plutus/blob/kwxm/machines-eval-builtins/metatheory/Scoped/CK.lagda.md. I am not sure how relevant it is. There is arity information everywhere - in the terms, in values and in the frames. And it is a bit obfuscated as the arity info is a proof that the arity is reasonable.

Reply:
Hmmm.  I'll think about that.  The version in `kwxm/alternative-ck` just uses frames though, and it has the advantage of being similar to the current version of the alternative CEK machine.  On the ogther hand, if this works maybe we could use it for the CEK machine as well.